### PR TITLE
Add langchain and streamlit

### PR DESCRIPTION
**Pull request by CatLuong0106**
Check this out. You don't have to merge the diarize.py code.

### CODE REVIEW

File: speaker_output.txt
Comment (menamerai) [inline speaker_output.txt:1]: We should probably not include this?

File: sen_analysis_test.py
Comment (menamerai) [inline sen_analysis_test.py:1]: Should we include this? Is there a sentimental analysis dev file somewhere?

Comment (crazymacaroni) [reply sen_analysis_test.py:1]: Keep it as a placeholder as I'm experimenting with it. If it's of no use, we can remove it later.

File: speaker_output.txt
Comment (crazymacaroni) [reply speaker_output.txt:1]: Yeah, I forgot to include it in .gitignore

File: langchain_dev.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+from langchain_core.output_parsers import StrOutputParser
+from langchain_core.runnables import RunnablePassthrough
+from langchain.chains import RetrievalQA
+from langchain.memory import ConversationBufferMemory
+from langchain.chains import ConversationalRetrievalChain
+
+
+os.environ["HUGGINGFACEHUB_API_TOKEN"] = get_token()
+
+def llm_init(repo_id: str = "google/gemma-7b") -> Any:
```
Comment (menamerai) [inline langchain_dev.py:27]: Should probably not default to gemma, it's gated so we'll need to submit a request to use it. We can use it regardless, but the default should be something else that isn't gated.

File: langchain_dev.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    # Set up a Hugging Face Endpoint for Gemma 2b model
+    llm = HuggingFaceEndpoint(
+    repo_id=repo_id, max_length=1024, temperature=0.1
```
Comment (menamerai) [inline langchain_dev.py:30]: Are these hyperparams a good idea?

File: langchain_dev.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        for speaker_info in json_obj: 
+            text = ""
+            text = f"[{speaker_info['start']} : {speaker_info['end']}] {speaker_info['speaker']}: {speaker_info['text']}\n"
+            f.write(text)
+
+
+def few_shot_prompts():
+    # TODO: Work this to figureout a way to add this to the conversational chain.
+    # Define examples that include user queries and AI's answers specific to Kaggle competitions
+    examples = [
```
Comment (menamerai) [inline langchain_dev.py:49]: What are these examples? Few shot prompts? How is this related to our tasks? I thought we had a summarization task?

File: langchain_dev.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    )
+
+    query = "Is participating in Kaggle competitions worth my time?"
+
+def load_and_chunk(input_path: str) -> Chroma:
+    with open(input_path) as f:
+        text = f.read()
+
+    text_splitter = RecursiveCharacterTextSplitter.from_tiktoken_encoder(
+        model_name="gpt-4",
```
Comment (menamerai) [inline langchain_dev.py:105]: Should this default to GPT-4?

File: langchain_dev.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        chunk_size=1000,
+        chunk_overlap=20,
+    )    
+    texts = text_splitter.split_text(text)
+    
+    embedding_function = SentenceTransformerEmbeddings(model_name="all-MiniLM-L6-v2")
```
Comment (menamerai) [inline langchain_dev.py:111]: What is this model?

File: langchain_dev.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        memory=memory,
+        condense_question_prompt=CUSTOM_QUESTION_PROMPT
+    )
+
+    return conversational_chain({"question": question})
+
+
+def main(): 
+    llm = llm_init()
+    json_path = Path(r"C:\Users\luong\callsentra\output\sample2.json")
```
Comment (menamerai) [inline langchain_dev.py:160]: Don't hardcode the path

Comment (crazymacaroni) [reply langchain_dev.py:160]: It doesn't matter right now only for testing. Your diarize.py also hardcoded audio path.

File: langchain_dev.py (hunk 2 above)
Comment (crazymacaroni) [reply langchain_dev.py:30]: Not sure we can always change hyperparameters later. I haven't played around with it enough

File: langchain_dev.py (hunk 3 above)
Comment (crazymacaroni) [reply langchain_dev.py:49]: I will update the new code now. But essentially it's for the RAG pipeline. I went to sleep before finishing that.

**Review by menamerai — CHANGES_REQUESTED** [review]
Should make the models in `langchain_dev` more consistent? I'm not super why you made the choices that you did?

File: langchain_dev.py (hunk 4 above)
Comment (crazymacaroni) [reply langchain_dev.py:105]: Yea, I think it's fine. It's a langchain thingy.

File: langchain_dev.py (hunk 5 above)
Comment (crazymacaroni) [reply langchain_dev.py:111]: THis is a sentence transformers used for embeddings.

**Comment by CatLuong0106** [comment]
I will make it more streamlined later. It's only for the demo right now. The biggest change is probably to switch to a different default model that's not gated since you need access for Gemma.

**Comment by menamerai** [comment]
Yeah just change the base model to something else. Just merge it when it's done. I think it looks really good.
